doozerlib/repos.py

```diff
@@ -176,6 +184,48 @@ def conf_section(self, repotype, arch=ARCH_X86_64, enabled=None, section_name=No
 
         return result
 
+    async def list_rpms(self, arch: str):
+        """ List all available rpms in this rpm repository
+        """
+        if arch not in self._valid_arches:
+            raise ValueError('{} is not a valid arch!')
+
+        # Attempt to get the cached result before actually querying the repo
+        cache_key = (self.name, arch)
+        with Repo._list_rpms_cache_lock:
+            cached_result = Repo._list_rpms_cache.get(cache_key)
+            if cached_result is not None:  # cache hit
+                return cached_result
+
+        # Cache miss. Will actually query the repo.
```

Comment:
Think the cache miss also should be inside the lock. Otherwise, there could be multiple queries running at the same time.

Reply:
Putting repoquery inside of the lock will prevent from multiple queries over different repos. There is a possibility to run multiple queries over the same repo, I think it is acceptable for now. Though there could still be a room to optimize this by using, such as, more fine grained locks, but I'd like to see how much time this will take in production.